qiskit/algorithms/minimum_eigensolvers/vqe.py

```diff
@@ -0,0 +1,380 @@
+    the Hamiltonian :math:`H` of a given system [1].
+
+    An instance of VQE requires defining two algorithmic sub-components: a trial state (a.k.a.
+    ansatz) which is a :class:`QuantumCircuit`, and one of the classical
+    :mod:`~qiskit.algorithms.optimizers`.
+
+    The ansatz is varied, via its set of parameters, by the optimizer, such that it works towards a
+    state, as determined by the parameters applied to the ansatz, that will result in the minimum
+    expectation value being measured of the input operator (Hamiltonian).
+
+    The optimizer can either be one of Qiskit's optimizers, such as
+    :class:`~qiskit.algorithms.optimizers.SPSA` or a callable with the following signature:
+
+    .. note::
+
+        The callable _must_ have the argument names ``fun, x0, jac, bounds`` as indicated
+        in the following code block.
+
+    .. code-block:: python
+
+        from qiskit.algorithms.optimizers import OptimizerResult
+
+        def my_minimizer(fun, x0, jac=None, bounds=None) -> OptimizerResult:
+            # Note that the callable *must* have these argument names!
+            # Args:
+            #     fun (callable): the function to minimize
+            #     x0 (np.ndarray): the initial point for the optimization
+            #     jac (callable, optional): the gradient of the objective function
+            #     bounds (list, optional): a list of tuples specifying the parameter bounds
+
+            result = OptimizerResult()
+            result.x = # optimal parameters
+            result.fun = # optimal function value
+            return result
+
+    The above signature also allows one to directly pass any SciPy minimizer, for instance as
+
+    .. code-block:: python
+
+        from functools import partial
+        from scipy.optimize import minimize
+
+        optimizer = partial(minimize, method="L-BFGS-B")
+
+    Attributes:
+        estimator: The estimator primitive to compute the expectation value of the circuits.
+        ansatz: The parameterized circuit used as an ansatz for the wave function.
+        optimizer: A classical optimizer to find the minimum energy. This can either be a
+            Qiskit :class:`.Optimizer` or a callable implementing the :class:`.Minimizer` protocol.
+        gradient: An optional gradient function or operator for the optimizer.
+        initial_point: An optional initial point (i.e. initial parameter values) for the optimizer.
+            If not provided, a random initial point with values in the interval :math:`[0, 2\pi]`
+            is used.
+
+    References:
+        [1] Peruzzo et al, "A variational eigenvalue solver on a quantum processor"
+            `arXiv:1304.3061 https://arxiv.org/abs/1304.3061>`_
+    """
+
+    def __init__(
+        self,
+        estimator: BaseEstimator,
+        ansatz: QuantumCircuit | None = None,
+        optimizer: Optimizer | Minimizer | None = None,
```

Comment:
We don't any forced keywords here, everything can be positional?

Reply:
I intend to change this. Please see [above](https://github.com/Qiskit/qiskit-terra/pull/8702#discussion_r972946302).